lib/vrl/stdlib/src/redact.rs

```diff
@@ -39,170 +58,185 @@ impl Function for Redact {
         let value = arguments.required("value");
 
         let filters = arguments
-            .optional_enum_list("filters", &Filter::all_str())?
-            .unwrap_or_default()
+            .required_array("filters")?
             .into_iter()
-            .map(|s| Filter::from_str(&s).expect("validated enum"))
-            .collect::<Vec<_>>();
-
-        let redactor = arguments
-            .optional_enum("redactor", &Redactor::all_str())?
-            .map(|s| Redactor::from_str(&s).expect("validated enum"))
-            .unwrap_or_default();
+            .map(|expr| {
+                expr.as_value()
+                    .ok_or(vrl::function::Error::ExpectedStaticExpression {
+                        keyword: "filters",
+                        expr,
+                    })
+            })
+            .map(|value| {
+                value.and_then(|value| {
+                    value.clone().try_into().map_err(|error| {
+                        vrl::function::Error::InvalidArgument {
+                            keyword: "filters",
+                            value,
+                            error,
+                        }
+                    })
+                })
+            })
+            .collect::<std::result::Result<Vec<Filter>, _>>()?;
 
-        let patterns = arguments.optional_array("patterns")?.map(Into::into);
+        let redactor = Redactor::Full;
 
         Ok(Box::new(RedactFn {
             value,
             filters,
             redactor,
-            patterns,
         }))
     }
 }
 
-// -----------------------------------------------------------------------------
+//-----------------------------------------------------------------------------
 
-#[derive(Debug)]
+#[derive(Clone, Debug)]
 struct RedactFn {
     value: Box<dyn Expression>,
     filters: Vec<Filter>,
     redactor: Redactor,
-    patterns: Option<Vec<Expr>>,
+}
+
+fn redact(value: Value, filters: &[Filter], redactor: &Redactor) -> Value {
+    match value {
+        Value::Bytes(bytes) => {
+            let input = String::from_utf8_lossy(&bytes);
+            let output = filters
+                .iter()
+                .fold(input, |input, filter| filter.redact(input, redactor));
+            Value::Bytes(output.into_owned().into())
+        }
+        Value::Array(values) => {
+            let values = values
+                .into_iter()
+                .map(|value| redact(value, filters, redactor))
+                .collect();
+            Value::Array(values)
+        }
+        Value::Object(map) => {
+            let map = map
+                .into_iter()
+                .map(|(key, value)| (key, redact(value, filters, redactor)))
+                .collect();
```

Comment:
Doing something like:

```rust
for (_, value) in map.iter_mut() {
  *value = redact(value, filters, redactor)
}
```

would save an allocation.

Reply:
👍 I tried this, but couldn't get it working. The first issue I ran into is that the recursive `redact` call needs a `vrl::Value` but we only have a `&mut` in the loop. I tried changing `redact` to take a `&'a vrl::Value` and return a `&'a vrl::Value` but then I couldn't assign to `*value` in the loop as `redact` returns a reference now:

```
119 |                 *value = redact(value, filters, redactor)
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `vrl::Value`, found `&vrl::Value`
```

I could clone, but I'm unsure if that'd actually end up saving any allocations then.